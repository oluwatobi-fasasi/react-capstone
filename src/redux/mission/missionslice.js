import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';

const initialState = {
  missions: [],
};

export const fetchMissionAsync = createAsyncThunk('mission/fetchMission', async () => {
  const response = await fetch('https://api.spacexdata.com/v3/missions');
  const data = await response.json();
  return data;
});

const missionSlice = createSlice({
  name: 'mission',
  initialState,
  reducers: {
    setMissions: (state, action) => ({
      ...state,
      missions: action.payload,
    }),
    reserveMission: (state, action) => {
      const missionId = action.payload;
      return {
        ...state,
        missions: state.missions.map((mission) => (mission.mission_id === missionId
          ? { ...mission, reserved: true } : mission)),
      };
    },

    cancelMission: (state, action) => {
      const missionId = action.payload;
      return {
        ...state,
        missions: state.missions.map((mission) => (mission.mission_id === missionId
          ? { ...mission, reserved: false } : mission)),
      };
    },
  },
});

export const selectAllMissions = (state) => state.mission.missions;

export const { setMissions, reserveMission, cancelMission } = missionSlice.actions;

export default missionSlice.reducer;
