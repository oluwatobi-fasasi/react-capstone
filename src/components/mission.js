import React from 'react';
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';
import { reserveMission, cancelMission } from '../redux/mission/missionslice';

const Mission = ({ mission }) => {
  const dispatch = useDispatch();

  const handleReserveMission = () => {
    dispatch(reserveMission(mission.mission_id));
  };

  const handleCancelMission = () => {
    dispatch(cancelMission(mission.mission_id));
  };

  return (
    <div className="mission-container">
      <div className="mission-info">
        <h2 className="miss-name">{mission.mission_name}</h2>
        <p className="miss-desc">{mission.description}</p>
      </div>
      <div className="mission-status">
        {mission.reserved ? (
          <span className="active-member">Active member</span>
        ) : (
          <span className="not-member">Not a member</span>
        )}
      </div>
      <div className="mission-buttons">
        {mission.reserved ? (
          <button type="submit" className="cancel-button" onClick={handleCancelMission}>
            Leave Mission
          </button>
        ) : (
          <button type="submit" className="join-button" onClick={handleReserveMission}>
            Join Mission
          </button>
        )}
      </div>
    </div>
  );
};

Mission.propTypes = {
  mission: PropTypes.shape({
    mission_id: PropTypes.string.isRequired,
    mission_name: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    reserved: PropTypes.bool.isRequired,
  }).isRequired,
};

export default Mission;
