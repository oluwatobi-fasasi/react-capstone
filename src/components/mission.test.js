import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import Mission from './mission';
import { reserveMission, cancelMission } from '../redux/mission/missionslice';

describe('Mission Component Tests', () => {
  const mockStore = configureStore();

  const mockMission = {
    mission_id: 'hello',
    mission_name: 'Test Mission',
    description: 'This is a test mission description.',
    reserved: false,
  };

  test('renders correctly when not reserved', () => {
    const store = mockStore({});
    const tree = renderer.create(
      <Provider store={store}>
        <Mission mission={mockMission} />
      </Provider>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  test('calls handleReserveMission on Join Mission button click', () => {
    const store = mockStore({});
    const component = renderer.create(
      <Provider store={store}>
        <Mission mission={mockMission} />
      </Provider>,
    );

    const joinButton = component.root.findByProps({ className: 'join-button' });
    joinButton.props.onClick();

    const actions = store.getActions();
    expect(actions).toEqual([reserveMission(mockMission.mission_id)]);
  });

  test('calls handleCancelMission on Leave Mission button click', () => {
    const reservedMission = {
      ...mockMission,
      reserved: true,
    };

    const store = mockStore({});
    const component = renderer.create(
      <Provider store={store}>
        <Mission mission={reservedMission} />
      </Provider>,
    );

    const leaveButton = component.root.findByProps({ className: 'cancel-button' });
    leaveButton.props.onClick();

    const actions = store.getActions();
    expect(actions).toEqual([cancelMission(reservedMission.mission_id)]);
  });

  test('renders correctly when reserved', () => {
    const reservedMission = {
      ...mockMission,
      reserved: true,
    };

    const store = mockStore({});
    const tree = renderer.create(
      <Provider store={store}>
        <Mission mission={reservedMission} />
      </Provider>,
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
