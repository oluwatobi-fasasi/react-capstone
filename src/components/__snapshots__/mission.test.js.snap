// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mission Component Tests renders correctly when not reserved 1`] = `
<div
  className="mission-container"
>
  <div
    className="mission-info"
  >
    <h2
      className="miss-name"
    >
      Test Mission
    </h2>
    <p
      className="miss-desc"
    >
      This is a test mission description.
    </p>
  </div>
  <div
    className="mission-status"
  >
    <span
      className="not-member"
    >
      Not a member
    </span>
  </div>
  <div
    className="mission-buttons"
  >
    <button
      className="join-button"
      onClick={[Function]}
      type="submit"
    >
      Join Mission
    </button>
  </div>
</div>
`;

exports[`Mission Component Tests renders correctly when reserved 1`] = `
<div
  className="mission-container"
>
  <div
    className="mission-info"
  >
    <h2
      className="miss-name"
    >
      Test Mission
    </h2>
    <p
      className="miss-desc"
    >
      This is a test mission description.
    </p>
  </div>
  <div
    className="mission-status"
  >
    <span
      className="active-member"
    >
      Active member
    </span>
  </div>
  <div
    className="mission-buttons"
  >
    <button
      className="cancel-button"
      onClick={[Function]}
      type="submit"
    >
      Leave Mission
    </button>
  </div>
</div>
`;
